### PR TITLE
[opentitantool] Add support for Ultradebug

### PR DESCRIPTION
**Pull request by cfrantz**
1. Create an interface for the UART (ultradebug ftdi interface C).
2. Create an interface for GPIOs (ultradebug ftdi interface B).
3. Create an interface for SPI (ultradebug ftdi interface B).
4. Create an MPSSE command executor to drive SPI transactions.

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

**Comment by imphil** [comment]
Do we have documentation in the tree outlining the ultradebug protocol?

**Comment by cfrantz** [comment]
> Do we have documentation in the tree outlining the ultradebug protocol?

There is no "ultradebug protocol" per se.  The ultradebug device is an FTDI FT4232H chip with a bit of extra glue logic for tristating certain pins, setting VREF and power control and monitoring.  The behavior of the extra glue logic is basically fixed - there's not any additional protocol to control that logic.

The code in this PR utilizes a pre-existing [libftdi interface crate](https://crates.io/crates/safe-ftdi) for communication with the FTDI chip and implements an [MPSSE](https://www.ftdichip.com/Documents/AppNotes/AN_135_MPSSE_Basics.pdf) protocol module to facilitate SPI transactions.

The ultradebug device is a Google-only device.  I'm adding support for it because I'm familiar with its operation, so it provides a baseline for me and a mechanism to start developing SPI support in opentitantool (ie: there will be a follow-on PR that implements the SPI Flash protocol).  Google's platforms-team also has quite a few ultradebugs deployed in various labs, so there might be a use case for them.

Longer-term, we're planning a new debug dongle for use with OpenTitan.  The firmware for that device will be open and its protocols will be documented.

File: sw/host/opentitanlib/Cargo.toml [hunk 1]
```diff
@@ -10,11 +10,16 @@ edition = "2018"
...
+thiserror = "1.0"
 lazy_static = "1.4.0"
 regex = "1"
 nix = "0.17.0"
 bitflags = "1.0"
 log = "0.4"
+# TODO(cfrantz): work with cr1901 to release 0.3.0.
+safe-ftdi = { git = "https://github.com/cr1901/safe-ftdi" }
 
+serde = {version="1", features=["serde_derive"]}
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/Cargo.toml:22]: Seems like something a formatter-tool should be fixing for me.

File: sw/host/opentitanlib/src/transport/ultradebug/mod.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+
+use log;
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/transport/ultradebug/mod.rs:7]: Removed.

File: sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+    DisableDivBy5,
+    InvalidCommand,
+}
+
+impl Command<'_, '_> {
+    /// Calculate the response length for this `Command`.
+    pub fn response_length(&self) -> usize {
+        match self {
+            Command::ReadData(_, buf) => buf.len(),
+            Command::WriteData(_, _) => 0,
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs:125]: Done.

File: sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+            Command::InvalidCommand => 0,
+        }
+    }
+
+    /// Extend a `Vec<u8>` with the low-level representation of this `Command`.
+    pub fn extend(&self, buf: &mut Vec<u8>) {
+        match self {
+            Command::ReadData(options, data) => {
+                buf.push(options.as_opcode());
+                buf.extend_from_slice(&((data.len() - 1) as u16).to_le_bytes());
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs:140]: Added an error condition for too-long buffers.

File: sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    }
+
+    /// Execute a slice of `commands` on the target FTDI device.
+    pub fn execute(&self, commands: &mut [Command]) -> Result<()> {
+        // Build up a transmit buffer from the slice of commands.
+        // The buffer contains both MPSSE opcodes and data.
+        // Calculate the size of the receive buffer needed.
+        let mut txbuf = Vec::new();
+        let mut rxlen = 0;
+        for command in commands.iter() {
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs:255]: If I write that, rust wants to move/consume `commands` into the for loop.  I need to call `iter_mut()` again on the slice later, so I'm leaving the explicit call to `iter`.  Is there a better way?

[the remark this answers is not in the document]
Comment (mcy) [reply sw/host/opentitanlib/src/transport/ultradebug/mpsse.rs:269]: There is not. =(

**Review by mcy — APPROVED** [review]

**Pull request merged**